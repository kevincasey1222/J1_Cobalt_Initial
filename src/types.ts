import { IntegrationInstanceConfig } from '@jupiterone/integration-sdk-core';

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * The provider API apiKeyAuth used to authenticate requests.
   */
  apiKeyAuth: string;

  /**
   * The provider API orgToken used to authenticate requests.
   */
  orgToken: string;
}
