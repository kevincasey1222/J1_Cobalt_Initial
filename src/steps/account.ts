import {
  createDirectRelationship,
  createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../types';

export const DATA_ACCOUNT_ENTITY = 'DATA_ACCOUNT_ENTITY';

export async function fetchAccountDetails({
  jobState,
  instance,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const name = `Cobalt - ${instance.name}`;
  const accountEntity = await jobState.addEntity(
    createIntegrationEntity({
      entityData: {
        source: {
          id: 'Cobalt',
          name: 'Cobalt Account',
        },
        assign: {
          _key: `cobalt-account:${instance.id}`,
          _type: 'cobalt_account',
          _class: 'Account',
          name: name,
          displayName: name,
        },
      },
    }),
  );
  await jobState.setData(DATA_ACCOUNT_ENTITY, accountEntity);

  const vendorEntity = await jobState.addEntity(
    createIntegrationEntity({
      entityData: {
        source: {
          id: 'Cobalt',
          name: 'Cobalt',
        },
        assign: {
          _key: `cobalt-vendor`,
          _type: 'cobalt_vendor',
          _class: 'Vendor',
          name: 'Cobalt',
          displayName: 'Cobalt',
          category: 'pentesters',
        },
      },
    }),
  );

  const serviceEntity = await jobState.addEntity(
    createIntegrationEntity({
      entityData: {
        source: {
          id: 'Cobalt-pentest-service',
          name: 'Cobalt pentest service',
        },
        assign: {
          _key: `cobalt-pentest-service`,
          _type: 'cobalt_service',
          _class: 'Service',
          name: 'Cobalt pentest service',
          displayName: 'Cobalt pentest service',
          category: ['pentesters'],
        },
      },
    }),
  );

  await jobState.addRelationship(
    createDirectRelationship({
      _class: RelationshipClass.PROVIDES,
      from: vendorEntity,
      to: serviceEntity,
    }),
  );

  await jobState.addRelationship(
    createDirectRelationship({
      _class: RelationshipClass.HAS,
      from: accountEntity,
      to: serviceEntity,
    }),
  );
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-account',
    name: 'Fetch Account Details',
    entities: [
      {
        resourceName: 'Cobalt Account',
        _type: 'cobalt_account',
        _class: 'Account',
      },
      {
        resourceName: 'Cobalt',
        _type: 'cobalt_vendor',
        _class: 'Vendor',
      },
      {
        resourceName: 'Cobalt pentest service',
        _type: 'cobalt_service',
        _class: 'Service',
      },
    ],
    relationships: [
      {
        _type: 'cobalt_account_has_service',
        _class: RelationshipClass.HAS,
        sourceType: 'cobalt_account',
        targetType: 'cobalt_service',
      },
      {
        _type: 'cobalt_vendor_provides_service',
        _class: RelationshipClass.PROVIDES,
        sourceType: 'cobalt_vendor',
        targetType: 'cobalt_service',
      },
    ],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
