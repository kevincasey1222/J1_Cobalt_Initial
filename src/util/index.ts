export function getVulnerabilityLink(description): string {
  //returns either "none", a CWE link at mitre.org, or a CVE link at nvd.nist.gov

  let webLink: string = 'none';
  const matches = description.match(/https?:\/\/(cwe.mitre.org).+html/);
  // Example: https://cwe.mitre.org/data/definitions/307.html
  if (matches) {
    webLink = matches[0];
  }
  const matches2 = description.match(/https?:\/\/(cve.mitre.org).*CVE[0-9_-]+/);
  // Example: https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-2138
  if (matches2) {
    // reformat URL to nvd.nist.gov URL, like https://nvd.nist.gov/vuln/detail/CVE-2021-2138
    const cveNumAry = matches2[0].match(/CVE[0-9_-]+/);
    webLink = `https://nvd.nist.gov/vuln/detail/${cveNumAry[0]}`;
  }
  const matches3 = description.match(/https?:\/\/(nvd.nist.gov).*CVE[0-9_-]+/);
  //Example: https://nvd.nist.gov/vuln/detail/CVE-2021-2138
  if (matches3) {
    webLink = matches3[0];
  }
  return webLink;
}

export function getVulnerabilityNumber(link): string {
  const cveMatch = link.match(/CVE[0-9_-]+/);
  //Example: CVE-2021-2138
  if (cveMatch) {
    return cveMatch[0];
  }
  const cweMatch = link.match(/[0-9_-]+\.html/);
  //Example: 307.html
  if (cweMatch) {
    return `CWE-${cweMatch[0].match(/[0-9_-]+/)[0]}`; //trim off the .html
  }
  return 'Unknown vulnerability';
}
