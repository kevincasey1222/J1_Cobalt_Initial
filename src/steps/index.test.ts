import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';

import { IntegrationConfig } from '../types';
import { fetchAccountDetails } from './account';
import { fetchFindings } from './findings';
import { fetchPentests } from './pentests';
import { fetchAssets } from './assets';
import { getVulnerabilityLink, getVulnerabilityNumber } from '../util';

//const DEFAULT_API_KEY = 'dummy-api-key';

const integrationConfig: IntegrationConfig = {
  apiKeyAuth:
    process.env.API_KEY_AUTH || 'L17FtgVqScP4tRsHcHGumU8eD7KZLHuqiMGtebir', //DEFAULT_API_KEY,
};

test('should collect data', async () => {
  const context = createMockStepExecutionContext<IntegrationConfig>({
    instanceConfig: integrationConfig,
  });

  // Simulates dependency graph execution.
  // See https://github.com/JupiterOne/sdk/issues/262.
  await fetchAccountDetails(context);
  await fetchAssets(context);
  await fetchPentests(context);
  await fetchFindings(context);

  // Review snapshot, failure is a regression
  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot();

  const accounts = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Account'),
  );
  expect(accounts.length).toBeGreaterThan(0);
  expect(accounts).toMatchGraphObjectSchema({
    _class: ['Account'],
    schema: {
      properties: {
        _type: { const: 'cobalt_account' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: [],
    },
  });

  const vendors = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Vendor'),
  );
  expect(vendors.length).toBeGreaterThan(0);
  expect(vendors).toMatchGraphObjectSchema({
    _class: ['Vendor'],
    schema: {
      properties: {
        _type: { const: 'cobalt_vendor' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: [],
    },
  });

  const services = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Service'),
  );
  expect(services.length).toBeGreaterThan(0);
  expect(services).toMatchGraphObjectSchema({
    _class: ['Service'],
    schema: {
      properties: {
        _type: { const: 'cobalt_service' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: [],
    },
  });

  const pentests = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Assessment'),
  );
  expect(pentests).toMatchGraphObjectSchema({
    _class: ['Assessment'],
    schema: {
      properties: {
        _type: { const: 'cobalt_pentest' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: [],
    },
  });

  const findings = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Finding'),
  );
  expect(findings).toMatchGraphObjectSchema({
    _class: ['Finding'],
    schema: {
      properties: {
        _type: { const: 'cobalt_finding' },
        pentestId: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['pentestId'],
    },
  });
});

describe('testing regex extraction of URLs from pentest descriptions', () => {
  test('recognize link and number of CWE', () => {
    const input =
      'Your weakness is basically https://cwe.mitre.org/data/definitions/307.html okay?';
    expect(getVulnerabilityLink(input)).toBe(
      'https://cwe.mitre.org/data/definitions/307.html',
    );
    expect(getVulnerabilityNumber(getVulnerabilityLink(input))).toBe('CWE-307');
  });

  test('recognize link and number of CVE from mitre', () => {
    const input =
      'Your vulnerability is basically https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-2138 okay?';
    expect(getVulnerabilityLink(input)).toBe(
      'https://nvd.nist.gov/vuln/detail/CVE-2021-2138', //yes, it's expected to change mitre to nvd
    );
    expect(getVulnerabilityNumber(getVulnerabilityLink(input))).toBe(
      'CVE-2021-2138',
    );
  });

  test('recognize link and number of CVE from nvd', () => {
    const input =
      'Your vulnerability is basically https://nvd.nist.gov/vuln/detail/CVE-2021-2138 okay?';
    expect(getVulnerabilityLink(input)).toBe(
      'https://nvd.nist.gov/vuln/detail/CVE-2021-2138',
    );
    expect(getVulnerabilityNumber(getVulnerabilityLink(input))).toBe(
      'CVE-2021-2138',
    );
  });
});
//more tests todo:
//build in recording, with appropriate redaction, take out real API key

//validate regex url scraper
//end of of word marker to regex for urls so we don't get multiple in one match
//validate bad token handling and refresh process
//consider - should be just discard a finding without a pentest? Right now throws error
//update snapshots when ready
