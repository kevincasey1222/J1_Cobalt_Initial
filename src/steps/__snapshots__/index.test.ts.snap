// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "cobalt-account:local-integration-instance",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "Cobalt",
            "name": "Cobalt Account",
          },
        },
      ],
      "_type": "cobalt_account",
      "createdOn": undefined,
      "displayName": "Cobalt - Local Integration",
      "id": "Cobalt",
      "name": "Cobalt - Local Integration",
    },
    Object {
      "_class": Array [
        "Vendor",
      ],
      "_key": "cobalt-vendor",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "Cobalt",
            "name": "Cobalt Vendor",
          },
        },
      ],
      "_type": "cobalt_vendor",
      "category": "security",
      "createdOn": undefined,
      "displayName": "Cobalt",
      "id": "Cobalt",
      "name": "Cobalt",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "cobalt-pentest-service",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "Cobalt-pentest-service",
            "name": "Cobalt pentest service",
          },
        },
      ],
      "_type": "cobalt_service",
      "category": Array [
        "security",
      ],
      "createdOn": undefined,
      "displayName": "Cobalt pentest service",
      "id": "Cobalt-pentest-service",
      "name": "Cobalt pentest service",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "as_9ctVo4X",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "resource": Object {
              "asset_type": "api",
              "description": "",
              "id": "as_9ctVo4X",
              "title": "Test API Asset",
            },
          },
        },
      ],
      "_type": "cobalt_asset",
      "assetType": "api",
      "attachments": undefined,
      "createdOn": undefined,
      "description": "",
      "displayName": "Test API Asset",
      "name": "Test API Asset",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "as_rPjSpXj",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "resource": Object {
              "asset_type": "web",
              "description": "First test Asset",
              "id": "as_rPjSpXj",
              "title": "Test Asset #1",
            },
          },
        },
      ],
      "_type": "cobalt_asset",
      "assetType": "web",
      "attachments": undefined,
      "createdOn": undefined,
      "description": "First test Asset",
      "displayName": "Test Asset #1",
      "name": "Test Asset #1",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "as_HrZ74Uh",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "resource": Object {
              "asset_type": "cloud_config",
              "description": "",
              "id": "as_HrZ74Uh",
              "title": "Test Cloud Config",
            },
          },
        },
      ],
      "_type": "cobalt_asset",
      "assetType": "cloud_config",
      "attachments": undefined,
      "createdOn": undefined,
      "description": "",
      "displayName": "Test Cloud Config",
      "name": "Test Cloud Config",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "as_ClGUrXE",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "resource": Object {
              "asset_type": "web",
              "description": "",
              "id": "as_ClGUrXE",
              "title": "Test Crazy Chars Asset @#$%@(~~\`\`",
            },
          },
        },
      ],
      "_type": "cobalt_asset",
      "assetType": "web",
      "attachments": undefined,
      "createdOn": undefined,
      "description": "",
      "displayName": "Test Crazy Chars Asset @#$%@(~~\`\`",
      "name": "Test Crazy Chars Asset @#$%@(~~\`\`",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "as_Vw8HhoR",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "resource": Object {
              "asset_type": "external_network",
              "description": "",
              "id": "as_Vw8HhoR",
              "title": "Test External Network Asset",
            },
          },
        },
      ],
      "_type": "cobalt_asset",
      "assetType": "external_network",
      "attachments": undefined,
      "createdOn": undefined,
      "description": "",
      "displayName": "Test External Network Asset",
      "name": "Test External Network Asset",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "as_O6qPyLc",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "resource": Object {
              "asset_type": "internal_network",
              "description": "",
              "id": "as_O6qPyLc",
              "title": "Test Internal Network Asset",
            },
          },
        },
      ],
      "_type": "cobalt_asset",
      "assetType": "internal_network",
      "attachments": undefined,
      "createdOn": undefined,
      "description": "",
      "displayName": "Test Internal Network Asset",
      "name": "Test Internal Network Asset",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "as_KmzMiKh",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "resource": Object {
              "asset_type": "mobile",
              "description": "Some description of a  mobile asset",
              "id": "as_KmzMiKh",
              "title": "Test Mobile Asset",
            },
          },
        },
      ],
      "_type": "cobalt_asset",
      "assetType": "mobile",
      "attachments": undefined,
      "createdOn": undefined,
      "description": "Some description of a  mobile asset",
      "displayName": "Test Mobile Asset",
      "name": "Test Mobile Asset",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "as_WFbWToK",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "resource": Object {
              "asset_type": "web_plus_api",
              "description": "",
              "id": "as_WFbWToK",
              "title": "Test Web+API Asset",
            },
          },
        },
      ],
      "_type": "cobalt_asset",
      "assetType": "web_plus_api",
      "attachments": undefined,
      "createdOn": undefined,
      "description": "",
      "displayName": "Test Web+API Asset",
      "name": "Test Web+API Asset",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "as_m8KEIsR",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "resource": Object {
              "asset_type": "web",
              "description": "",
              "id": "as_m8KEIsR",
              "title": "Test Web Asset",
            },
          },
        },
      ],
      "_type": "cobalt_asset",
      "assetType": "web",
      "attachments": undefined,
      "createdOn": undefined,
      "description": "",
      "displayName": "Test Web Asset",
      "name": "Test Web Asset",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "as_qpShMvk",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "resource": Object {
              "asset_type": "web_plus_external_network",
              "description": "",
              "id": "as_qpShMvk",
              "title": "Test Web + External Network Asset",
            },
          },
        },
      ],
      "_type": "cobalt_asset",
      "assetType": "web_plus_external_network",
      "attachments": undefined,
      "createdOn": undefined,
      "description": "",
      "displayName": "Test Web + External Network Asset",
      "name": "Test Web + External Network Asset",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "as_MyGvYsm",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "resource": Object {
              "asset_type": "web_plus_mobile",
              "description": "
Как вы сегодня 
Test of non English chars

كيف حالك اليوم",
              "id": "as_MyGvYsm",
              "title": "Test Web+Mobile Asset",
            },
          },
        },
      ],
      "_type": "cobalt_asset",
      "assetType": "web_plus_mobile",
      "attachments": undefined,
      "createdOn": undefined,
      "description": "
Как вы сегодня 
Test of non English chars

كيف حالك اليوم",
      "displayName": "Test Web+Mobile Asset",
      "name": "Test Web+Mobile Asset",
    },
    Object {
      "_class": Array [
        "Assessment",
      ],
      "_key": "pt_gm4lmL1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "resource": Object {
              "asset_id": null,
              "end_date": "Feb 22 2021",
              "id": "pt_gm4lmL1",
              "methodology": null,
              "objectives": "",
              "platform_tags": Array [
                "ReactJS",
                "RoR",
              ],
              "start_date": "Feb 8 2021",
              "state": "live",
              "tag": "#PT5734",
              "targets": Array [
                "Cobalt.io",
              ],
              "title": "Test Asset #1 - February 2021",
            },
          },
        },
      ],
      "_type": "cobalt_pentest",
      "assetId": null,
      "category": "Penetration Testing",
      "createdOn": undefined,
      "displayName": "Test Asset #1 - February 2021",
      "endDate": "Feb 22 2021",
      "internal": false,
      "methodology": null,
      "name": "Test Asset #1 - February 2021",
      "objectives": "",
      "platformTags": "[
  \\"ReactJS\\",
  \\"RoR\\"
]",
      "startDate": "Feb 8 2021",
      "state": "live",
      "summary": "Test Asset #1 - February 2021",
      "tag": "#PT5734",
      "targets": "[
  \\"Cobalt.io\\"
]",
    },
    Object {
      "_class": Array [
        "Finding",
      ],
      "_key": "vu_dJ749NP",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "resource": Object {
              "affected_targets": Array [
                "cobalt.io",
              ],
              "asset_id": "as_rPjSpXj",
              "description": "Most security problems are weaknesses in an application that result from a broken or missing security control (authentication, access control, input validation, etc...). By contrast, business logic vulnerabilities are ways of using the legitimate processing flow of an application in a way that results in a negative consequence to the organization.
",
              "id": "vu_dJ749NP",
              "impact": 0,
              "labels": Array [],
              "likelihood": 0,
              "log": Array [],
              "pentest_id": "pt_gm4lmL1",
              "proof_of_concept": null,
              "severity": "low",
              "state": "need_fix",
              "suggested_fix": "something",
              "tag": "#PT5734_2",
              "title": "Business logic",
              "type_category": "Business Logic",
            },
          },
        },
      ],
      "_type": "cobalt_finding",
      "affectedTargets": "[
  \\"cobalt.io\\"
]",
      "assetId": "as_rPjSpXj",
      "category": "Penetration Testing",
      "createdOn": undefined,
      "description": "Most security problems are weaknesses in an application that result from a broken or missing security control (authentication, access control, input validation, etc...). By contrast, business logic vulnerabilities are ways of using the legitimate processing flow of an application in a way that results in a negative consequence to the organization.
",
      "displayName": "Business logic",
      "impact": "0",
      "labels": "[]",
      "likelihood": 0,
      "log": "[]",
      "name": "Business logic",
      "numericSeverity": 0,
      "open": true,
      "pentestId": "pt_gm4lmL1",
      "prerequisites": undefined,
      "proofOfConcept": null,
      "severity": "0",
      "state": "need_fix",
      "suggestedFix": "something",
      "tag": "#PT5734_2",
      "typeCategory": "Business Logic",
    },
    Object {
      "_class": Array [
        "Finding",
      ],
      "_key": "vu_gIGDRPl",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "resource": Object {
              "affected_targets": Array [
                "Cobalt.io",
              ],
              "asset_id": "as_rPjSpXj",
              "description": "Cross-Site Scripting (XSS) attacks are a type of injection, in which malicious scripts are injected into otherwise benign and trusted websites. XSS attacks occur when an attacker uses a web application to send malicious code, generally in the form of a browser side script, to a different end user. Flaws that allow these attacks to succeed are quite widespread and occur anywhere a web application uses input from a user within the output it generates without validating or encoding it.

An attacker can use XSS to send a malicious script to an unsuspecting user. The end user’s browser has no way to know that the script should not be trusted, and will execute the script. Because it thinks the script came from a trusted source, the malicious script can access any cookies, session tokens, or other sensitive information retained by the browser and used with that site. These scripts can even rewrite the content of the HTML page.

Stored attacks are those where the injected script is permanently stored on the target servers, such as in a database, in a message forum, visitor log, comment field, etc. The victim then retrieves the malicious script from the server when it requests the stored information. Stored XSS is also sometimes referred to as Persistent or Type-I XSS.

",
              "id": "vu_gIGDRPl",
              "impact": 0,
              "labels": Array [],
              "likelihood": 0,
              "log": Array [],
              "pentest_id": "pt_gm4lmL1",
              "proof_of_concept": null,
              "severity": "high",
              "state": "need_fix",
              "suggested_fix": "1. Always treat all user input as untrusted data. 2. Never insert untrusted data except in allowed locations. 3. Always input or output-encode all data coming into or out of the application. 4. Always whitelist allowed characters and seldom use blacklisting of characters except in certain use cases. 5. Always use a well-known and security encoding API for input and output encoding such as the \`OWASP ESAPI\`. 6. Never try to write input and output encoders unless absolutely necessary. Chances are that someone has already written a good one. 7. Never use the DOM function \`innerHtml\` and instead use the functions \`innerText\` and \`textContent\` to prevent against DOM-based XSS. 8. As a best practice, consider using the \`HTTPOnly\` flag on cookies that are session tokens or sensitive tokens. 9. As a best practice, consider implementing \`Content Security Policy\` to protect against XSS and other injection type attacks. 10. As a best practice, consider using an auto-escaping templating system. 11. As a best practice, consider using the \`X-XSS-Protection\` response header.",
              "tag": "#PT5734_1",
              "title": "First findings",
              "type_category": "Validation, Sanitization and Encoding",
            },
          },
        },
      ],
      "_type": "cobalt_finding",
      "affectedTargets": "[
  \\"Cobalt.io\\"
]",
      "assetId": "as_rPjSpXj",
      "category": "Penetration Testing",
      "createdOn": undefined,
      "description": "Cross-Site Scripting (XSS) attacks are a type of injection, in which malicious scripts are injected into otherwise benign and trusted websites. XSS attacks occur when an attacker uses a web application to send malicious code, generally in the form of a browser side script, to a different end user. Flaws that allow these attacks to succeed are quite widespread and occur anywhere a web application uses input from a user within the output it generates without validating or encoding it.

An attacker can use XSS to send a malicious script to an unsuspecting user. The end user’s browser has no way to know that the script should not be trusted, and will execute the script. Because it thinks the script came from a trusted source, the malicious script can access any cookies, session tokens, or other sensitive information retained by the browser and used with that site. These scripts can even rewrite the content of the HTML page.

Stored attacks are those where the injected script is permanently stored on the target servers, such as in a database, in a message forum, visitor log, comment field, etc. The victim then retrieves the malicious script from the server when it requests the stored information. Stored XSS is also sometimes referred to as Persistent or Type-I XSS.

",
      "displayName": "First findings",
      "impact": "0",
      "labels": "[]",
      "likelihood": 0,
      "log": "[]",
      "name": "First findings",
      "numericSeverity": 0,
      "open": true,
      "pentestId": "pt_gm4lmL1",
      "prerequisites": undefined,
      "proofOfConcept": null,
      "severity": "0",
      "state": "need_fix",
      "suggestedFix": "1. Always treat all user input as untrusted data. 2. Never insert untrusted data except in allowed locations. 3. Always input or output-encode all data coming into or out of the application. 4. Always whitelist allowed characters and seldom use blacklisting of characters except in certain use cases. 5. Always use a well-known and security encoding API for input and output encoding such as the \`OWASP ESAPI\`. 6. Never try to write input and output encoders unless absolutely necessary. Chances are that someone has already written a good one. 7. Never use the DOM function \`innerHtml\` and instead use the functions \`innerText\` and \`textContent\` to prevent against DOM-based XSS. 8. As a best practice, consider using the \`HTTPOnly\` flag on cookies that are session tokens or sensitive tokens. 9. As a best practice, consider implementing \`Content Security Policy\` to protect against XSS and other injection type attacks. 10. As a best practice, consider using an auto-escaping templating system. 11. As a best practice, consider using the \`X-XSS-Protection\` response header.",
      "tag": "#PT5734_1",
      "typeCategory": "Validation, Sanitization and Encoding",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "PROVIDES",
      "_fromEntityKey": "cobalt-vendor",
      "_key": "cobalt-vendor|provides|cobalt-pentest-service",
      "_toEntityKey": "cobalt-pentest-service",
      "_type": "cobalt_vendor_provides_service",
      "displayName": "PROVIDES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "cobalt-account:local-integration-instance",
      "_key": "cobalt-account:local-integration-instance|has|cobalt-pentest-service",
      "_toEntityKey": "cobalt-pentest-service",
      "_type": "cobalt_account_has_service",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "cobalt-account:local-integration-instance",
      "_key": "cobalt-account:local-integration-instance|has|as_9ctVo4X",
      "_toEntityKey": "as_9ctVo4X",
      "_type": "cobalt_account_has_asset",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "cobalt-account:local-integration-instance",
      "_key": "cobalt-account:local-integration-instance|has|as_rPjSpXj",
      "_toEntityKey": "as_rPjSpXj",
      "_type": "cobalt_account_has_asset",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "cobalt-account:local-integration-instance",
      "_key": "cobalt-account:local-integration-instance|has|as_HrZ74Uh",
      "_toEntityKey": "as_HrZ74Uh",
      "_type": "cobalt_account_has_asset",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "cobalt-account:local-integration-instance",
      "_key": "cobalt-account:local-integration-instance|has|as_ClGUrXE",
      "_toEntityKey": "as_ClGUrXE",
      "_type": "cobalt_account_has_asset",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "cobalt-account:local-integration-instance",
      "_key": "cobalt-account:local-integration-instance|has|as_Vw8HhoR",
      "_toEntityKey": "as_Vw8HhoR",
      "_type": "cobalt_account_has_asset",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "cobalt-account:local-integration-instance",
      "_key": "cobalt-account:local-integration-instance|has|as_O6qPyLc",
      "_toEntityKey": "as_O6qPyLc",
      "_type": "cobalt_account_has_asset",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "cobalt-account:local-integration-instance",
      "_key": "cobalt-account:local-integration-instance|has|as_KmzMiKh",
      "_toEntityKey": "as_KmzMiKh",
      "_type": "cobalt_account_has_asset",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "cobalt-account:local-integration-instance",
      "_key": "cobalt-account:local-integration-instance|has|as_WFbWToK",
      "_toEntityKey": "as_WFbWToK",
      "_type": "cobalt_account_has_asset",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "cobalt-account:local-integration-instance",
      "_key": "cobalt-account:local-integration-instance|has|as_m8KEIsR",
      "_toEntityKey": "as_m8KEIsR",
      "_type": "cobalt_account_has_asset",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "cobalt-account:local-integration-instance",
      "_key": "cobalt-account:local-integration-instance|has|as_qpShMvk",
      "_toEntityKey": "as_qpShMvk",
      "_type": "cobalt_account_has_asset",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "cobalt-account:local-integration-instance",
      "_key": "cobalt-account:local-integration-instance|has|as_MyGvYsm",
      "_toEntityKey": "as_MyGvYsm",
      "_type": "cobalt_account_has_asset",
      "displayName": "HAS",
    },
    Object {
      "_class": "PERFORMED",
      "_fromEntityKey": "cobalt-vendor",
      "_key": "cobalt-vendor|performed|pt_gm4lmL1",
      "_toEntityKey": "pt_gm4lmL1",
      "_type": "cobalt_vendor_performed_pentest",
      "displayName": "PERFORMED",
    },
    Object {
      "_class": "PERFORMED",
      "_fromEntityKey": "cobalt-pentest-service",
      "_key": "cobalt-pentest-service|performed|pt_gm4lmL1",
      "_toEntityKey": "pt_gm4lmL1",
      "_type": "cobalt_service_performed_pentest",
      "displayName": "PERFORMED",
    },
    Object {
      "_class": "IDENTIFIED",
      "_fromEntityKey": "pt_gm4lmL1",
      "_key": "pt_gm4lmL1|identified|vu_dJ749NP",
      "_toEntityKey": "vu_dJ749NP",
      "_type": "cobalt_pentest_identified_finding",
      "displayName": "IDENTIFIED",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "as_rPjSpXj",
      "_key": "as_rPjSpXj|has|vu_dJ749NP",
      "_toEntityKey": "vu_dJ749NP",
      "_type": "cobalt_asset_has_finding",
      "displayName": "HAS",
    },
    Object {
      "_class": "IDENTIFIED",
      "_fromEntityKey": "pt_gm4lmL1",
      "_key": "pt_gm4lmL1|identified|vu_gIGDRPl",
      "_toEntityKey": "vu_gIGDRPl",
      "_type": "cobalt_pentest_identified_finding",
      "displayName": "IDENTIFIED",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "as_rPjSpXj",
      "_key": "as_rPjSpXj|has|vu_gIGDRPl",
      "_toEntityKey": "vu_gIGDRPl",
      "_type": "cobalt_asset_has_finding",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "cobalt_account",
    "cobalt_vendor",
    "cobalt_service",
    "cobalt_vendor_provides_service",
    "cobalt_account_has_service",
    "cobalt_asset",
    "cobalt_account_has_asset",
    "cobalt_pentest",
    "cobalt_vendor_performed_pentest",
    "cobalt_service_performed_pentest",
    "cobalt_finding",
    "cobalt_pentest_identified_finding",
    "cobalt_asset_has_finding",
  ],
  "numCollectedEntities": 17,
  "numCollectedRelationships": 19,
}
`;
